### ISSUE TITLE
route/mount should also accept middlewares

### ISSUE THREAD

**Issue by tcurdt** [honojs/hono#2794]
### What is the feature you are proposing?

I had this

```
app
  .route('/foo/', PageFoo)
  .get('/protected', authMiddleware, (ctx) => ctx.text('protected', 200))
```

and then wanted to also protect `/foo/` by applying the `authMiddleware`. I assume this would work:

```
app
  .route('/foo/', authMiddleware, PageFoo)
  .get('/protected', authMiddleware, (ctx) => ctx.text('protected', 200))
```

but it seems that's not the case.

IMO `route` and `mount` (what's the difference?) should both accept middlewares.

Right now this causes

```
88 |   route(path, app) {
89 |     const subApp = this.basePath(path);
90 |     if (!app) {
91 |       return subApp;
92 |     }
93 |     app.routes.map((r) => {
         ^
TypeError: undefined is not an object (evaluating 'app.routes.map')
```

**Comment by nakasyou** [comment]
`app.route` receives other Hono instances, in contrast `app.mount` receives fetch handler like `(req: Request) => Promise<Response> | Response`.

@tcurdt, which content is `PageFoo`, handler, Hono instance or middleware?

**Comment by tcurdt** [comment]
It's an instance

`export const PageFoo = new Hono()`

That's why I assumed that this should work.

```
app
  .route('/foo/', authMiddleware, PageFoo)
  .get('/protected', authMiddleware, (ctx) => ctx.text('protected', 200))
```

What's the difference between `app.mount` and `app.all` then?

**Comment by tcurdt** [comment]
Of course there is (what I consider) a work around. Inside `PageFoo` I can `.use(authMiddleware)` but from an overview and API perspective it would be nice to also allow for middlewares as this `app.route(path, [middlewares], app)`

**Comment by nakasyou** [comment]
@tcurdt 
In `app.mount`, you can mount fetch handler like `(req: Request) => Promise<Response> | Response` from other frameworks.
In contrast, `app.all` adds handler. It's like `app.get` that accepts all methods.

Could you write code like that?:
```ts
app
  .route('/foo', PageFoo.use('*', authMiddleware))
  .get('/protected', authMiddleware, (ctx) => ctx.text('protected', 200))
```

**Comment by tcurdt** [comment]
I just tried

```
.route('/foo/', PageFoo.use('*', MiddlewareAuth).use('*', MiddlewareAuthRedirect))
.route('/foo/', PageFoo.use(MiddlewareAuth).use(MiddlewareAuthRedirect))
```

but surprisingly neither worked.
Maybe the order is be wrong? This passing style means the `PageFoo.use()` is coming after the `PageFoo.get()`.

From an API point of view it would also be a little odd to have `get` and `route` use different style of parameters IMO.

**Comment by nakasyou** [comment]
@tcurdt, thank you for comment.
In different idea, 
```ts
app
  .use('/foo/*', authMiddleware)
  .route('/foo', PageFoo)
  .get('/protected', authMiddleware, (ctx) => ctx.text('protected', 200))
```

**Comment by tcurdt** [comment]
That should work. One could probably separate this out then into.

```
app
  .use('/foo/*', authMiddleware)
  .use('/protected/*', authMiddleware)
  .route('/foo', PageFoo)
  .get('/protected', (ctx) => ctx.text('protected', 200))

```

But it does not feel like a great solution. Duplication and adding (suffix) matchers for every request.
I would probably rather put the middleware inside `PageFoo`.

But it's a shame. It would make it nice and clear if one could pass it to `app.route`. Would be my preferred way.

**Comment by EdamAme-x** [comment]
The `.route` takes the **Hono** instance and adds it to the routing. It is *type-safe* and can also use **RPC**.  
`.mount` routes simple request handlers or routers from other libraries in a *non-typesafe* manner.  
`.mount` accepts middleware, while `.route` only accepts non-Middleware handlers. If you want to include middleware, you can use `.mount("base path here", appIncludeMiddeware.fetch)`, which is not type-safe...

**Comment by tcurdt** [comment]
Thanks for the deeper explanation, @EdamAme-x. Much appreciated.

As an `hono` API user that distinction really is not that clear though. At least not to me.
Even with that explanation I am not sure why the duality of the two is actually needed.

Why not just have `route`?

**Comment by EdamAme-x** [comment]
Hi @tcurdt .
I am not sure if the following explanation is correct.
If without typing by `typescript`, `.route` could be merged into `.mount`. However, typing is required.
By its nature, `.mount` can only pass functions of the following types.
```ts
(request: Request, ...args: any[]) => Response | Promise<Response>
```
This is necessary for integration with other library routers.
`.route` accepts only Hono instances for accurate typing.
There are two distinctions here.

**Comment by tcurdt** [comment]
Hm. Maybe my TS foo is not strong enough yet. But it should be a very light layer to adapt between the two. 🤔
Anyway. Feels a little strange.
Thanks for trying to explain!

**Comment by EdamAme-x** [comment]
I don't know if this is the case either.
What do you think? @yusukebe

**Comment by yusukebe** [comment]
Hi @tcurdt 

Indeed, as you said, if the `app.route()` and `app.mount()` can receive middleware, it will be useful in some cases. However, we don't have a plan to implement it because the implementation and type definition will be complex. I think it is simple that `app.route()` only accepts an instance of Hono, and `app.mount()` only accepts a fetch handler.

**Comment by tcurdt** [comment]
> because the implementation and type definition will be complex

Just curious - what makes it more complex than `get` and friends? shouldn't it follow the same pattern?

**Comment by yusukebe** [comment]
> Just curious - what makes it more complex than `get` and friends? shouldn't it follow the same pattern?

Yes, it's the same pattern. But it will be like an overload, so we need to write the conditional branches. This may be a few lines, but increasing the file size is critical for the core of `hono`.
